File: patterns/04_decorator/javascript/triumphator.js
// ============================================================
//  DECORATOR PATTERN — The Triumphator's Dressing Room
//  Run: node triumphator.js
// ============================================================

class BareSoldier {
  describe()    { return "👤 Tiro (Recruit) — tunica, caligae sandals"; }
  combatPower() { return 5; }
  status()      { return "Recruit"; }
}
class SoldierDecorator {
  constructor(soldier) { this.soldier = soldier; }
  describe()    { return this.soldier.describe(); }
  combatPower() { return this.soldier.combatPower(); }
  status()      { return this.soldier.status(); }
}
class LoricaSegmentata extends SoldierDecorator {
  describe()    { return this.soldier.describe() + "\n              + 🛡  Lorica Segmentata"; }
  combatPower() { return this.soldier.combatPower() + 30; }
  status()      { return `Armoured ${this.soldier.status()}`; }
}
class GallicHelmet extends SoldierDecorator {
  describe()    { return this.soldier.describe() + "\n              + ⛑  Gallic Helmet"; }
  combatPower() { return this.soldier.combatPower() + 15; }
}
class PaludamentumCloak extends SoldierDecorator {
  describe()    { return this.soldier.describe() + "\n              + 🟥 Paludamentum Cloak"; }
  combatPower() { return this.soldier.combatPower() + 20; }
  status()      { return `General ${this.soldier.status()}`; }
}
class LaurelWreath extends SoldierDecorator {
  describe()    { return this.soldier.describe() + "\n              + 🌿 LAUREL WREATH OF TRIUMPH!"; }
  combatPower() { return this.soldier.combatPower() + 50; }
  status()      { return "TRIUMPHATOR"; }
  shout()       { return "IO TRIUMPHE! IO TRIUMPHE! IO TRIUMPHE!"; }
}

const show = s => {
  console.log(`  Status : ${s.status()}`);
  console.log(`  Outfit : ${s.describe()}`);
  console.log(`  Power  : ${s.combatPower()} points`);
};

console.log("╔═══════════════════════════════════════════════╗");
console.log("║   DECORATOR PATTERN — JavaScript              ║");
console.log("╚═══════════════════════════════════════════════╝\n");

let s = new BareSoldier();
console.log("── STEP 1:"); show(s);
s = new LoricaSegmentata(s);
console.log("\n── STEP 2: + Lorica"); show(s);
s = new GallicHelmet(s);
console.log("\n── STEP 3: + Helmet"); show(s);
s = new PaludamentumCloak(s);
console.log("\n── STEP 4: + General's Cloak"); show(s);
const t = new LaurelWreath(s);
console.log("\n── STEP 5: TRIUMPH!"); show(t);
console.log(`\n  🎺 ${t.shout()}`);
console.log('\n"Miles ornatur, non mutatur!"');

File: patterns/04_decorator/csharp/Triumphator.cs
// ============================================================
//  DECORATOR PATTERN — The Triumphator's Dressing Room
//  Run: dotnet script Triumphator.cs
// ============================================================
using System;

interface ISoldier { string Describe(); int CombatPower(); string Status(); }

class BareSoldier : ISoldier {
    public string Describe()    => "👤 Tiro (Recruit) — tunica, caligae sandals";
    public int    CombatPower() => 5;
    public string Status()      => "Recruit";
}
abstract class SoldierDecorator : ISoldier {
    protected ISoldier Soldier;
    protected SoldierDecorator(ISoldier s) { Soldier = s; }
    public virtual string Describe()    => Soldier.Describe();
    public virtual int    CombatPower() => Soldier.CombatPower();
    public virtual string Status()      => Soldier.Status();
}
class LoricaSegmentata : SoldierDecorator {
    public LoricaSegmentata(ISoldier s) : base(s) {}
    public override string Describe()    => Soldier.Describe() + "\n              + 🛡  Lorica Segmentata";
    public override int    CombatPower() => Soldier.CombatPower() + 30;
    public override string Status()      => $"Armoured {Soldier.Status()}";
}
class GallicHelmet : SoldierDecorator {
    public GallicHelmet(ISoldier s) : base(s) {}
    public override string Describe()    => Soldier.Describe() + "\n              + ⛑  Gallic Helmet";
    public override int    CombatPower() => Soldier.CombatPower() + 15;
}
class PaludamentumCloak : SoldierDecorator {
    public PaludamentumCloak(ISoldier s) : base(s) {}
    public override string Describe()    => Soldier.Describe() + "\n              + 🟥 Paludamentum Cloak";
    public override int    CombatPower() => Soldier.CombatPower() + 20;
    public override string Status()      => $"General {Soldier.Status()}";
}
class LaurelWreath : SoldierDecorator {
    public LaurelWreath(ISoldier s) : base(s) {}
    public override string Describe()    => Soldier.Describe() + "\n              + 🌿 LAUREL WREATH OF TRIUMPH!";
    public override int    CombatPower() => Soldier.CombatPower() + 50;
    public override string Status()      => "TRIUMPHATOR";
    public string Shout()                => "IO TRIUMPHE! IO TRIUMPHE! IO TRIUMPHE!";
}

static void Show(ISoldier s) {
    Console.WriteLine($"  Status : {s.Status()}");
    Console.WriteLine($"  Outfit : {s.Describe()}");
    Console.WriteLine($"  Power  : {s.CombatPower()} points");
}

Console.WriteLine("╔═══════════════════════════════════════════════╗");
Console.WriteLine("║   DECORATOR PATTERN — C#                      ║");
Console.WriteLine("╚═══════════════════════════════════════════════╝\n");

ISoldier s = new BareSoldier();
Console.WriteLine("── STEP 1:"); Show(s);
s = new LoricaSegmentata(s); Console.WriteLine("\n── STEP 2: + Lorica"); Show(s);
s = new GallicHelmet(s);     Console.WriteLine("\n── STEP 3: + Helmet"); Show(s);
s = new PaludamentumCloak(s);Console.WriteLine("\n── STEP 4: + General's Cloak"); Show(s);
var t = new LaurelWreath(s); Console.WriteLine("\n── STEP 5: TRIUMPH!"); Show(t);
Console.WriteLine($"\n  🎺 {t.Shout()}");
Console.WriteLine("\n\"Miles ornatur, non mutatur!\"");
